src/borg/testsuite/key.py

```diff
@@ -232,3 +248,225 @@ def test_passphrase_new_retries(self, monkeypatch):
+
+        # enough space in reservation
+        manager.ensure_reservation(12)
+        enc_cipher.expect_no_reset_and_advance(0x2000 + 20 + 12)
+        assert self.cache_nonce() == "0000000000002034"
+        assert self.repository.next_free == 0x2034
+
+        # just barely enough space in reservation
+        manager.ensure_reservation(20)
+        enc_cipher.expect_no_reset_and_advance(0x2000 + 20 + 12 + 20)
+        assert self.cache_nonce() == "0000000000002034"
+        assert self.repository.next_free == 0x2034
+
+        # no space in reservation
+        manager.ensure_reservation(16)
+        enc_cipher.expect_no_reset_and_advance(0x2000 + 20 + 12 + 20 + 16)
+        assert self.cache_nonce() == "0000000000002064"
+        assert self.repository.next_free == 0x2064
+
+        # spans reservation boundary
+        manager.ensure_reservation(64)
+        enc_cipher.expect_no_reset_and_advance(0x2000 + 20 + 12 + 20 + 16 + 64)
+        assert self.cache_nonce() == "00000000000020c4"
+        assert self.repository.next_free == 0x20c4
+
+    def test_sync_nonce(self, monkeypatch):
+        monkeypatch.setattr(borg.key, 'NONCE_SPACE_RESERVATION', 32)
+
+        enc_cipher = self.MockEncCipher(0x2000)
+        self.repository = self.MockRepository()
+        self.repository.next_free = 0x2000
+        self.set_cache_nonce("0000000000002000")
+
+        manager = NonceManager(self.repository, enc_cipher, 0x2000)
+        manager.ensure_reservation(20)
+        enc_cipher.expect_iv_and_advance(0x2000, 0x2000 + 20)
+
+        assert self.cache_nonce() == "0000000000002034"
+        assert self.repository.next_free == 0x2034
+
+    def test_server_just_upgraded(self, monkeypatch):
+        monkeypatch.setattr(borg.key, 'NONCE_SPACE_RESERVATION', 32)
+
+        enc_cipher = self.MockEncCipher(0x2000)
+        self.repository = self.MockRepository()
+        self.repository.next_free = None
+        self.set_cache_nonce("0000000000002000")
+
+        manager = NonceManager(self.repository, enc_cipher, 0x2000)
+        manager.ensure_reservation(20)
+        enc_cipher.expect_iv_and_advance(0x2000, 0x2000 + 20)
+
+        assert self.cache_nonce() == "0000000000002034"
+        assert self.repository.next_free == 0x2034
+
+    def test_transaction_abort_no_cache(self, monkeypatch):
+        monkeypatch.setattr(borg.key, 'NONCE_SPACE_RESERVATION', 32)
+
+        enc_cipher = self.MockEncCipher(0x1000)
+        self.repository = self.MockRepository()
+        self.repository.next_free = 0x2000
+
+        manager = NonceManager(self.repository, enc_cipher, 0x2000)
+        manager.ensure_reservation(20)
+        enc_cipher.expect_iv_and_advance(0x2000, 0x2000 + 20)
+
+        assert self.cache_nonce() == "0000000000002034"
+        assert self.repository.next_free == 0x2034
+
+    def test_transaction_abort_old_server(self, monkeypatch):
+        monkeypatch.setattr(borg.key, 'NONCE_SPACE_RESERVATION', 32)
```

Comment:
i don't mean having to convert 20 to 0x14.
if mean using 32 dec (== 20 hex) and 20 dec.